### DISCUSSION TITLE
Jlink device not showing up as a /dev device

### DISCUSSION THREAD

**Discussion by OriMiller** [bitcraze/discussions#912] [Q&A]
In May of last year we ordered 1 AI deck (order #5908). We were able Flash the Wi-Fi streaming example over the radio no problem. At the time we only had that one deck and unfortunately, during a test flight, the drone crashed and what we assume is the antenna fell off. From that point forward, we were no longer able to connect to the Wi-Fi network that was being set up by the AI deck. Our solution was to buy more AI decks. So last month we purchased two more (order # 7168). As soon as they arrived, we attempted to flash the Wi-Fi streaming example. Unfortunately, all we got were errors. Upon looking at your tutorial guides we found AI decks less than revision D did not come from the factory with the proper firmware and needed to be manually flashed after the fact. When we looked at all the AI decks together, we found that the first AI deck we bought in 2022 was revision D and the AI decks that we bought last month were all revision C1. This let us to your tutorial about using JTAG JLink device to flash them from a Linux computer. We ordered a Segger edu mini Jlink device and have been attempting to flash these AI decks to no avail. We've looked through tons of bit craze forums and Segger forums, and are just running into issues with getting the JLink device to show up as /dev/ttyUSB0 or being able to figure out which /dev it’s being set as for that matter. Any help would be greatly appreciated.

**Comment by tobbeanton** [comment]
How much have you investigated the edu mini Jlink? I'm not hundred percent sure but I think I've heard there are some limitations to it so it can not work with the GAP8 or perhaps it was the ESP32. Do you have some insights [@whoenig](https://github.com/whoenig)?

**Comment by BNoordam** [comment]
Are there any updates regarding this problem? I have run into exactly the same problem with the j-link EDU mini, and doubt if it is supposed to show up as a /dev device. What I do know is that this j-link does not have a FTDI interface and therefore doesn't show up as ttyUSB0. In discussion #852 a J-link (8.08.00 J-LINK BASE CLASSIC i guess) is used to build and flash the bootloader which pops up as a ttyACM0 device. How they managed this. or if this is automatically the case with this J-link I have not been able to discover. Would be very interested in further information on this, or to know if a workaround is possible with the docker command. I have seen some openOCD scripts that operate only with the .cfg file of the programmer, so without calling /dev/tty.

**Comment by Janamejay-Sharma** [comment]
Hi all,

I work with Ori, and I successfully flashed the bootloader on the GAP8 module using a J-Link EDU Mini on Ubuntu 20.04.

The official flashing guide provides the following command:
`docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'`

We need to make a few changes to this.

## Select the correct .cfg file

If you're using any J-Link device, make sure you change the .cfg file to `interface/jlink.cfg`.

## J-Link Device Path

@BNoordam is correct in that the J-Link EDU Mini does not show up as `/dev/ttyUSB0`. Here is how to find out the correct path:
1. Plug in the J-Link EDU Mini to your PC's USB port.
2. Run `lsusb`. Find the Segger J-Link in the list and take note of its Bus & Device number. For me, it read `Bus 001 Device 010`, so the required path on Ubuntu was `/dev/bus/usb/001/010`.

## Corrected Command
The final, correct command in my case was:
`docker run --rm -it -v $PWD:/module/ --device /dev/bus/usb/001/010 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/jlink.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'`

## Pitfalls To Avoid
#### Connection Order:
It's never stated explicitly, but the connection steps are:
1. Connect the 10-pin cable to the AI deck's GAP8 JTAG header. Note how the cable faces downwards.
!<img alt="AI Deck GAP8 JTAG" width="400" src="https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/images/ai-deck-jtag-connecting.png"/>

2. Connect the 10-pin cable to the J-Link EDU Mini. Ensure the cable faces the chip in the middle.

3. Power on the Crazyflie, and *then* connect the micro USB cable to your PC.

#### J-Link Driver
Make sure you have the J-Link driver installed. If you're on Ubuntu, I recommend installing it using the .deb package: https://www.segger.com/downloads/jlink/. **Restart** your computer after installing the driver.

Hope this helps!

**Reply by knmcguire** [reply]
thanks for this excessive guide! @BNoordam I hope this helps you too.

@Janamejay-Sharma if you feel that the [documentation of the aideck](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/) is lacking, you can perhaps do a pull request to fix the documentation? The source is in the doc folder of   https://github.com/bitcraze/aideck-gap8-examples

**Reply by zimka** [reply]
I also used J-Link Edu to flash AIDeck bootloader from VM. In my case it worked correctly with
`docker run --rm -it -v $PWD:/module/ -v /dev/bus/usb:/dev/bus/usb --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/jlink.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'`

**Comment by BNoordam** [comment]
Thanks indeed for your clear guide @Janamejay-Sharma! My jlink does indeed pop up in /dev/bus/usb/. However, when implementing this in the docker command I still receive the same error as before. I'm using Ubuntu 22.04, have installed the jlink driver, and followed the instructions on getting started with the AI-deck, including the right .cfg file and the pitfalls mentioned above. The output is as follows: 

` ~/aideck-gap8-bootloader$ docker run --rm -it -v $PWD:/module/ --device /dev/bus/usb/001/009 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/jlink.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/; make all image flash'
gapy --target=ai_deck --platform=board --work-dir=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS run --image --binary=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader
gapy --target=ai_deck --platform=board --work-dir=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS run --flash --binary=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader
make: Nothing to be done for 'image'.
gapy --target=ai_deck --platform=board --work-dir=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS run --flash --force --binary=/module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/bootloader
Flashing image with command:
gap8-openocd -d0 -c "gdb_port disabled; telnet_port disabled; tcl_port disabled" -c "script interface/jlink.cfg; script target/gap8revb.tcl; script tcl/flash_image.tcl; script tcl/jtag_boot.tcl; gap_flash_raw_hyper /module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img 75808 /gap_sdk/utils/openocd_tools; exit;"
Open On-Chip Debugger 0.10.0+dev-gb84d97ec4 (2022-04-20-14:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
debug_level: 0
DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
GAP8 INIT TARGET
GAP8 examine target
Init jtag
Initialising GAP8 JTAG TAP
begining flash session (hyperflash)

load flasher to L2 memory
Loading binary through JTAG
Error: Sending data to device timed out.
Error: transport_write() failed: timeout occurred.
Error: jaylink_jtag_io() failed: timeout occurred. `

these 3 errors keep repeating. The output ends as follows:

912 bytes written at address 0x1b000004 Traceback (most recent call last): File "/gap_sdk/utils/gapy/gapy", line 162, in <module> main() File "/gap_sdk/utils/gapy/gapy", line 152, in main operationFunc(args, config, system) File "/gap_sdk/utils/gapy/run.py", line 184, in operationFunc raise RuntimeError('Runner has failed with value: %d' % status) RuntimeError: Runner has failed with value: -1 make: *** [/gap_sdk/rtos/freeRTOS/vendors/gwt/rules/freeRTOS_rules.mk:404: flash] Error 1 
I think my Jlink still isn't recognised. @knmcguire do you know what else could be the problem?

**Reply by knmcguire** [reply]
Hi! it seems that your jlink is being recognized now but something else is the problem, namely that it can't flash the gap8. The jtag cable that @Janamejay-Sharma has mentioned, I've seen those that are in their reverse... so lint coming in from the top (annoying I know...).

Another thing, is that the jtag cable can be broken, also a common issue. But before that, perhaps make sure to take the full power of the crazyflie by deconnecting the battery and everything to make sure that the gap8 is not in a locked state and restart it. Then connect and reconnect the cable again. Just some general tips that might work... 

keep us posted!

**Reply by BNoordam** [reply]
Thanks for the tips. So far it hasn't worked yet. I will try to find another jtag cable coming Monday and give an update if this has worked. It still seems strange to me though that I receive the exact same output now as I did before when I didn't have the right /dev location in the docker command.

**Reply by knmcguire** [reply]
Hmm oke. Then perhaps the location wasn't the issue. It could also be that openocd was already able to detect your device as it is no longer necessary to indicate the location as long as your .cfg is selected (I'm heavily speculating here). 🤔 

The thing is, that something has started and if your device wasn't detected you wouldn't even see this `beginning flash session (hyperflash)`. Also we didn't see your original output problem before and since 3 people are now in this discussion it might be that there are different causes to consider. It makes it difficult for us to help out here.

Perhaps, if you are still having problems, to start up a new discussion would maybe be better, referring to this one?

**Reply by Janamejay-Sharma** [reply]
@BNoordam, since your error code includes `Error: transport_write() failed: timeout occurred.`, try using a different USB port, or try adding the line `adapter_khz 1000` to your jlink.cfg file. Source: https://dzone.com/articles/eclipse-jtag-debugging-the-esp32-with-a-segger-j-l

**Reply by BNoordam** [reply]
@Janamejay-Sharma Thanks again for the response. I have tried a different pc with ubuntu 20.04 and different USB ports. I haven't changed anything, but receive a different output now;
`DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
GAP8 INIT TARGET
Error: JTAG scan chain interrogation failed: all ones
Error: Check JTAG interface, timings, target power, etc.
Error: Trying to use configured scan chain anyway...
Error: gap8.cpu: IR capture error; saw 0x0f not 0x01
GAP8 examine target
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
Error: Burst read failed
can't read halt state
Error: Couldn't read the CPU state
Error: Burst read failed
can't read halt state
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
Error: Burst read failed
can't read halt state
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
Error: Burst read failed
can't read halt state
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
Error: Burst read failed
can't read halt state
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
Error: Burst read failed
can't read halt state
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
Error: Could not re-establish communication with target
Error: Error while calling gap8_is_cpu_running
Polling target gap8.cpu failed, trying to reexamine
GAP8 examine target
Error: Burst read failed
can't read halt state
Error: Couldn't read the CPU state
Examination failed, GDB will be halted. Polling again in 100ms
`
Is it possible to alter the jlink.cfg file when using docker? As the jlink.cfg file is in the openOCD repository.

**Comment by Lorite** [comment]
If you are facing issues with the Jlink not being detected, I found out that my J-Link was in an incorrect mode. Open JLinkConfigExe and connect the board. Then, double click it. A configuration window will open. Select "SEGGER" on the last option "USB Driver (Windows)". Mine was on "WinUSB" before, which didn't work. Also, I am using Linux, not Windows, so the naming might be wrong... Feel free to check the image below. As it was initially pointed out here, the J-Link should work: https://github.com/orgs/bitcraze/discussions/912.
![Screenshot From 2025-07-02 17-20-55](https://github.com/user-attachments/assets/415f6718-8bac-4106-b1a6-0da5b3979e27)

**Reply by ToveRumar** [reply]
Hi @Lorite,

Thanks a lot for the clear explanation!
It is a bit surprising that changing the driver was necessary in the first place, however it likely means others might run into the same issue as well.

We really appreciate contributions like this from the community!
